File: 2353-design-a-food-rating-system/2353-design-a-food-rating-system.java
import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;

class FoodRatings {
	public record Entry(Integer rating, String food) implements Comparable<Entry> {

		@Override
		public int compareTo(Entry other) {
			if (this.rating.equals(other.rating)) {
				return food.compareTo(other.food);
			}
			return -(this.rating - other.rating);
		}
	}
	public record LookupVal(String cuisine, Entry entry) {
	}

	private Map<String, TreeSet<Entry>> q = new HashMap();
	private Map<String, LookupVal> lookup = new HashMap();

	public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
		int n = foods.length;

		for (int i = 0; i < n; i++) {
			var cuisine = cuisines[i];
			var food = foods[i];
			var rating = ratings[i];

			var entry = new Entry(rating, food);

			q.putIfAbsent(cuisine, new TreeSet<>());
			q.get(cuisine).add(entry);
			lookup.put(food, new LookupVal(cuisine, entry));
		}
	}

	public void changeRating(String food, int newRating) {
		var lookupval = lookup.get(food);
		q.get(lookupval.cuisine()).remove(lookupval.entry());

		var newEntry = new Entry(newRating, food);
		q.get(lookupval.cuisine()).add(newEntry);
		lookup.put(food, new LookupVal(lookupval.cuisine, newEntry));
	}

	public String highestRated(String cuisine) {
		return this.q.get(cuisine).first().food();
	}
}

/**
 * Your FoodRatings object will be instantiated and called as such: FoodRatings obj = new
 * FoodRatings(foods, cuisines, ratings); obj.changeRating(food,newRating); String param_2 =
 * obj.highestRated(cuisine);
 */

File: 2353-design-a-food-rating-system/2353-design-a-food-rating-system.py
from collections import defaultdict
from typing import List

from sortedcontainers import SortedList


class FoodRatings:
    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        """
        food -> entry (rating)
        cuisine -> entry
        """
        self.lookup = dict()
        self.q = defaultdict(
            lambda: SortedList(key=lambda it: (-it[0], it))
        )  # cuisine -> (rating, food) queue
        for food, cuisine, rating in zip(foods, cuisines, ratings):
            self.q[cuisine].add((rating, food))
            self.lookup[food] = (cuisine, (rating, food))
        return

    def changeRating(self, food: str, new_rating: int) -> None:
        cuisine, old_key = self.lookup[food]
        self.q[cuisine].remove(old_key)

        new_key = (new_rating, food)
        self.q[cuisine].add(new_key)

        self.lookup[food] = (cuisine, new_key)
        return

    def highestRated(self, cuisine: str) -> str:
        return self.q[cuisine][0][1]


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)